src/storage/src/source/postgres/replication.rs

```diff
@@ -173,97 +204,98 @@ pub(crate) fn render<G: Scope<Timestamp = MzOffset>>(
+                            let commit_lsn = MzOffset::from(begin.final_lsn());
+
+                            let mut tx = pin!(extract_transaction(
+                                stream.by_ref(),
+                                commit_lsn,
+                                &table_info,
+                                &connection_config,
+                                &metrics,
+                                &connection.publication,
+                                &mut errored
+                            ));
+
+                            trace!(%id, "timely-{worker_id} extracting transaction at {commit_lsn}");
+                            while let Some((oid, event, diff)) = tx.try_next().await? {
+                                if !table_info.contains_key(&oid) {
+                                    continue;
                                 }
-                                new_upper = tx_lsn + 1;
-                                if container.len() > max_capacity {
-                                    data_output.give_container(data_cap, &mut container).await;
-                                    upper_cap.downgrade(&new_upper);
-                                    data_cap.downgrade(&new_upper);
+                                let data = (oid, event);
+                                if let Some((rewind_cap, req)) = rewinds.get(&oid) {
+                                    if commit_lsn <= req.snapshot_lsn {
+                                        let update = (data.clone(), MzOffset::from(0), -diff);
+                                        data_output.give(rewind_cap, update).await;
+                                    }
                                 }
+                                assert!(new_upper <= commit_lsn, "new_upper={} tx_lsn={}", new_upper, commit_lsn);
+                                container.push((data, commit_lsn, diff));
                             }
-                            Some(Ok(_)) => return Err(TransientError::BareTransactionEvent),
-                            Some(Err(_)) => {
-                                return Err(stream.as_mut().next().await.unwrap().unwrap_err())
+                            new_upper = commit_lsn + 1;
+                            if container.len() > max_capacity {
+                                data_output.give_container(data_cap, &mut container).await;
+                                upper_cap.downgrade(&new_upper);
+                                data_cap.downgrade(&new_upper);
                             }
-                            None => break,
-                        }
+                        },
+                        _ => return Err(TransientError::BareTransactionEvent),
+                    }
+                    Ok(PrimaryKeepAlive(keepalive)) => {
+                        trace!(%id, "timely-{worker_id} received keepalive lsn={}", keepalive.wal_end());
+                        new_upper = std::cmp::max(new_upper, keepalive.wal_end().into());
```

Comment:
I need to do some digging to validate this, but my current understanding is that this is equivalent to having received a txn whose commit LSN is this value, so we should add 1--this value is likely to become the `confirmed_flush_lsn`, and we don't want to theoretically ingest a txn with this LSN.

```suggestion
                        new_upper = std::cmp::max(new_upper, MzOffset::from(keepalive.wal_end()) + 1);
```

I'll note that I think this point is somewhat moot because this message explicitly states that we won't see a txn at this time unless we already have––but for stepping through the logic of the program and reverse-engineering the semantics, I think it's helpful.

Reply:
My understanding is that the `sentPtr` already contains the `+1` in it so we shouldn't do it twice here. I had the `+1` in the begining and I observed panics in the local version of this PR which had aggressive asserts on every message on the modified postgres that spammed the stream with keepalives.

I did some extra digging now and also found [this](https://github.com/postgres/postgres/blob/REL_16_STABLE/src/include/access/xlogreader.h#L207) in the source code of postgres which says `EndRecPtr` is `end + 1` and [this](https://github.com/postgres/postgres/blob/REL_16_STABLE/src/backend/replication/walsender.c#L3083) which sets the `sentPtr` to `EndRecPtr`, which is what eventually makes into the keepalive value.

Reply:
Convincing links––I hadn't gotten to look at the PG source code yet. Thanks for being ahead of that.